ee/kernel/src/timer_alarm.c

```diff
@@ -0,0 +1,264 @@
+ *
+ * Stops the alarm. If the alarm was not armed it does nothing.
+ * When an alarm triggers it is automatically disabled, so user must
+ * call iStartTimerAlarm to re-enable it.
+ */
+void StopTimerAlarm(struct timer_alarm_t *alarm) {
+    u32 oldintr = DIntr();
+    iStopTimerAlarm(alarm);
+    if (oldintr)
+        EIntr();
+}
+
+/** Same as StartTimerAlarm but for IRQ handlers.
+ *
+ */
+void iStartTimerAlarm(struct timer_alarm_t *alarm) {
+    // Bail if the alarm is already set!
+    if (alarm->scheduled_time)
+        return;
+
+    // Calculate clock scheduling time
+    u64 sched_time = ps2_clock() * 256 + alarm->timer_cycles;
+    alarm->scheduled_time = sched_time;
+
+    if (!event_queue) {
+        // No alarm was set
+        event_queue = alarm;
+        alarm->next = alarm->prev = NULL;
+        Timer2Resched(alarm);
+    }
+    else if (event_queue->scheduled_time > alarm->scheduled_time) {
+        // This alarm is the next alarm to trigger
+        alarm->next = event_queue;
+        alarm->prev = NULL;
+        event_queue->prev = alarm;
+        event_queue = alarm;
+        Timer2Resched(alarm);
+    }
+    else {
+        // Just queue it at the right place (keep the list sorted)
+        struct timer_alarm_t *it = event_queue;
+        for (; it != NULL; it = it->next) {
+            if (!it->next || it->next->scheduled_time > alarm->scheduled_time) {
+                alarm->next = it->next;
+                alarm->prev = it;
+                if (it->next)
+                    it->next->prev = alarm;
+                it->next = alarm;
+                break;
+            }
+        }
+        // No need to update the alarm at all
+    }
+}
+
+/** Starts (arms) an alarm.
+ *
+ * @param alarm Alarm object to be started.
+ *
+ * Starts the alarm. If the alarm was already started, it does nothing.
+ */
+void StartTimerAlarm(struct timer_alarm_t *alarm) {
+    u32 oldintr = DIntr();
+    iStartTimerAlarm(alarm);
+    if (oldintr)
+        EIntr();
+}
+
+/** Sets up alarm paramenters.
+ *
+ * @param alarm Alarm object to be updated
+ * @param clock_cycles Number of bus cycles when the alarm will trigger
+ * @param user_callback Function pointer to call when the alarm triggers (in IRQ)
+ * @param user_arg Argument to pass to user_callback
+ *
+ * Starts the alarm. If the alarm was already started, it does nothing.
+ */
+void SetTimerAlarm(struct timer_alarm_t *alarm, u64 clock_cycles, timer_alarm_callback_t user_callback, void *user_arg) {
+    // Do not set alarms that are armed.
+    if (alarm->scheduled_time)
+        return;
+
+    // ps2_clock is based on a 576kHz timer
+    alarm->timer_cycles = clock_cycles;
+    alarm->callback = user_callback;
+    alarm->usr_arg = user_arg;
+}
+
+static void _wake_sema_cb(struct timer_alarm_t *alarm, void *userptr) {
+    (void)alarm;
+
+    s32 sema_id = *(s32*)userptr;
+    iSignalSema(sema_id);
+}
+
+void ThreadWaitClock(u64 clock_cycles) {
```

Comment:
how does it behave with `0` as input?

Reply:
For consistency we guarantee that all alarms will trigger. This is important if there are many alarms close in time, we do not want to "miss" one of them. Thats why in the code, when we calculate the timer cycles we do this:

```
s64 wait_time = (s64)alarm->scheduled_time - (s64)(ps2_clock() * 256);
u64 alarm_clocks = wait_time <= 0 ? 1 : wait_time;```

Which means: if the alarm is in the past or present (negative delay or zero), we schedule 1 cycle, so that the timer will be triggered immediately. Again, for consistency. With zero of very small values the alarm triggers as soon as the StartTimerAlarm exits since the timer runs out almost immediately.

Reply:
That's a really good approach.